### PR TITLE
Fix servlet metric names

### PR DESCRIPTION
**Pull request by JorikSchellekens**

### CODE REVIEW

File: synapse/http/server.py [hunk 1]
```diff
@@ -328,7 +327,7 @@ def _get_handler_for_request(self, request):
             m = path_entry.pattern.match(request.path.decode("ascii"))
             if m:
                 # We found a match!
-                return path_entry.callback, m.groupdict()
+                return path_entry.callback, path_entry.servlet_classname, m.groupdict()
```
Comment (richvdh) [inline synapse/http/server.py]: need to update the docstring

File: synapse/http/server.py [hunk 2]
```diff
@@ -328,7 +327,7 @@ def _get_handler_for_request(self, request):
...
 
         # Huh. No one wanted to handle that? Fiiiiiine. Send 400.
         return _unrecognised_request_handler, {}
```
Comment (richvdh) [inline synapse/http/server.py]: this needs an update too

File: synapse/http/server.py [hunk 3]
```diff
@@ -255,12 +257,12 @@ def __init__(self, hs, canonical_json=True):
         self.path_regexs = {}
         self.hs = hs
 
-    def register_paths(self, method, path_patterns, callback):
+    def register_paths(self, method, path_patterns, callback, servlet_classname):
```
Comment (richvdh) [inline synapse/http/server.py]: could probably do with taking the opportunity to add a docstring here. What does this method do, what do its params do, and what are their expected types?

File: changelog.d/5734.bugfix [hunk 4]
```diff
@@ -0,0 +1 @@
+Fix servlet metric names.
```
Comment (richvdh) [inline changelog.d/5734.bugfix]: could you enlarge on this a little? Are the metrics in question prometheus metrics or some other kind? How are they broken? When was the brokenness introduced?

File: synapse/http/server.py [hunk 5]
```diff
@@ -275,12 +277,9 @@ def render(self, request):
...
             path.
         """
-        callback, group_dict = self._get_handler_for_request(request)
+        callback, servlet_classname, group_dict = self._get_handler_for_request(request)
 
-        servlet_instance = getattr(callback, "__self__", None)
-        if servlet_instance is not None:
-            servlet_classname = servlet_instance.__class__.__name__
-        else:
+        if servlet_classname is None:
```
Comment (richvdh) [inline synapse/http/server.py]: when would it be None?

Comment (JorikSchellekens) [reply synapse/http/server.py]: I'm just defending against the event that someone passes None for the classname. We still need a unique name to reference it in prometheus.

Comment (JorikSchellekens) [reply synapse/http/server.py]: I've added a comment to that effect.

File: changelog.d/5734.bugfix (hunk 4 above)
Comment (JorikSchellekens) [reply changelog.d/5734.bugfix]: How's that?

File: changelog.d/5734.bugfix [hunk 6]
```diff
@@ -0,0 +1,3 @@
+The decorated servlets were producing cryptic names in prometheus after the
```
Comment (richvdh) [inline changelog.d/5734.bugfix]: sorry, this is probably too wordy now :). Also the release number is probably more relevant to most readers than the

How about:

> Fix a regression introduced in v1.2.0rc1 which led to incorrect labels on some prometheus metrics.

File: synapse/http/server.py [hunk 7]
```diff
@@ -311,7 +327,8 @@ def _get_handler_for_request(self, request):
             request (twisted.web.http.Request):
 
         Returns:
-            Tuple[Callable, dict[unicode, unicode]]: callback method, and the
+            Tuple[Callable, str, dict[unicode, unicode]]: callback method, the
+                prefferred name to use for that method in prometheus and the
```
Comment (richvdh) [inline synapse/http/server.py]:
```suggestion
                label to use for that method in prometheus metrics, and the
```

File: synapse/http/server.py (hunk 5 above)
Comment (richvdh) [reply synapse/http/server.py]: If we need to defend against that, I think I'd rather see it done in `register_paths`: however, I don't think it's a thing we need to defend against?

Comment (JorikSchellekens) [reply synapse/http/server.py]: To be honest I just threw it in because I had the same confusion from line 281 in the old code. Why was there the test `if servlet_instance is not None:`? I'll remove it.

File: changelog.d/5734.bugfix (hunk 6 above)
Comment (JorikSchellekens) [reply changelog.d/5734.bugfix]: Perfect

File: synapse/http/server.py (hunk 5 above)
Comment (JorikSchellekens) [reply synapse/http/server.py]: It's not necessary. I've removed it.

Comment (richvdh) [reply synapse/http/server.py]: > Why was there the test `if servlet_instance is not None`:

because sometimes the `callback` would be a regular function rather than an instance method (for example: `_options_handler`), so would have no `__self__` attribute. That's no longer a concern.

Comment (JorikSchellekens) [reply synapse/http/server.py]: Aha, that makes sense
